internal/upgradecheck/header.go

```diff
@@ -0,0 +1,174 @@
+func generateHeader(log logr.Logger, version string, crclient crclient.Client,
+	ctx context.Context, cfg *rest.Config, isOpenShift bool) *clientUpgradeData {
+
+	var ensureDeploymentIDErr, getManagedClustersErr, getServerVersionErr error
+	var pgoClustersTotal int
+
+	upgradeInfo := &clientUpgradeData{
+		PGOVersion:  version,
+		IsOpenShift: isOpenShift,
+	}
+
+	deploymentID, ensureDeploymentIDErr = ensureDeploymentID(crclient, ctx)
+	// ensureDeploymentID always returns some ID, so log the error and continue
+	if ensureDeploymentIDErr != nil {
+		log.Error(
+			fmt.Errorf("%w", ensureDeploymentIDErr),
+			"Error in getting deploymentID, using backup",
+		)
+	}
+	upgradeInfo.DeploymentID = deploymentID
+	pgoClustersTotal, getManagedClustersErr = getManagedClusters(crclient, ctx)
+	// getManagedClusters always returns some count, so log the error and continue
+	if getManagedClustersErr != nil {
+		log.Error(
+			fmt.Errorf("%w", getManagedClustersErr),
+			"Error in getting total number of clusters",
+		)
+	}
+	upgradeInfo.PGOClustersTotal = pgoClustersTotal
+
+	serverVersion, getServerVersionErr := getServerVersion(cfg)
+	// getServerVersion always returns some version.Info, so log the error and continue
+	if getServerVersionErr != nil {
+		log.Error(
+			fmt.Errorf("%w", getServerVersionErr),
+			"Error in getting server version",
+		)
+	}
+	upgradeInfo.KubernetesEnv = serverVersion
+	return upgradeInfo
+}
+
+// ensureDeploymentID checks if the UUID exists in memory or in a ConfigMap
+// If no UUID exists, ensureDeploymentID creates one and saves it in memory/as a ConfigMap
+func ensureDeploymentID(crclient crclient.Client, ctx context.Context) (uuid.UUID, error) {
+	var err error
+	var configMapUUID uuid.UUID
+	cm := &corev1.ConfigMap{}
+
+	// If there is no deploymentID in memory, generate one for possible use
+	if deploymentID == uuid.Nil {
+		deploymentID, err = uuid.NewUUID()
+	}
+
+	// Ensure a configmap exists with a UUID
+	// If it doesn't exist, create it with the in-memory ID
+	// If it exists, use it as the in-memory ID
+	// PGO only ever creates the ConfigMap -- if it exists, it is the source of truth for the ID
+	if err == nil {
+		if err = crclient.Get(ctx, naming.AsObjectKey(
+			naming.UpgradeCheckConfigMap(),
+		), cm); err != nil {
+			if apierrors.IsNotFound(err) {
+				cm = &corev1.ConfigMap{ObjectMeta: naming.UpgradeCheckConfigMap()}
+				cm.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("ConfigMap"))
+				cm.Data = map[string]string{
+					"deployment_id": deploymentID.String(),
+				}
+				err = errors.WithStack(crclient.Create(ctx, cm))
+				if err != nil {
+					return deploymentID, err
+				}
+
+			} else {
+				// If we encounter any error except NotFound, just go ahead with the generated ID
+				// and skip the ConfigMap
+				return deploymentID, err
+			}
+		}
+	}
+
+	if err == nil {
+		configMapUUID, err = uuid.Parse(cm.Data["deployment_id"])
```

Comment:
:wrench: One issue I'm seeing here is that if the `deployment_id` is deleted from the ConifgMap (but the ConfigMap itself still exists), we'll never update the ConfigMap with what we have in memory.  Can we use a Server-Side Apply in this function instead of using `Create()`?

I do realize our `apply()` function as used for SSA is in the `postgrescluster` package.  Maybe we move it elsewhere so that it can also be used here as well?

I also wouldn't be against just creating `apply()` and `patch()` funcitons in this package that serve the specific use case of the upgrade checker (e.g. this thing only deals with ConfigMap's, so no need to worry about the workarounds for StatefulSets, Jobs and Services in the current `apply()` function).

The benefit of a SSA is that we'll always being applying our intent, which means we'll always set the `deployment_id` if its missing.  And if we need to add additional fields to the ConfigMap in the future, we can simply update our intent, and those changes will be patched (and overall this aligns with our use of SSA for creating/patching resources throughout other parts of PGO v5).

Reply:
That makes sense to me, with the caveat that we don't want to overwrite our deployment_id unless it's... not a valid UUID? Like:
CM doesn't exist -- create
CM exists, has deployment id UUID -- leave
CM exists, deployment id doesn't exist or is malformed -- update

Reply:
And for checking that some string is a valid UUID, I'd probably use that google/uuid package

Reply:
Ahh yes!  I do see we would need the Google package to check the UUID.

Also, I concur with the following :+1: 

> CM doesn't exist -- create
> CM exists, has deployment id UUID -- leave
> CM exists, deployment id doesn't exist or is malformed -- update